querier.go

```diff
@@ -791,31 +795,50 @@ func (it *chunkSeriesIterator) Seek(t int64) (ok bool) {
 		t = it.mint
 	}
 
-	for ; it.chunks[it.i].MaxTime < t; it.i++ {
-		if it.i == len(it.chunks)-1 {
-			return false
-		}
+	// Return early if already past t.
+	if t0, _ := it.cur.At(); t0 >= t {
+		return t0 <= it.maxt
 	}
 
-	it.cur = it.chunks[it.i].Chunk.Iterator()
-	if len(it.intervals) > 0 {
-		it.cur = &deletedIterator{it: it.cur, intervals: it.intervals}
-	}
+	iCur := it.i
+	for ; it.i < len(it.chunks); it.i++ {
+		// Skip chunks with MaxTime < t.
+		if it.chunks[it.i].MaxTime < t {
+			continue
+		}
 
-	for it.cur.Next() {
-		t0, _ := it.cur.At()
-		if t0 >= t {
-			return true
+		// Don't reset the iterator unless we've moved on to a different chunk.
+		if it.i != iCur {
+			it.cur = it.chunks[it.i].Chunk.Iterator()
+			if len(it.intervals) > 0 {
+				it.cur = &deletedIterator{it: it.cur, intervals: it.intervals}
+			}
+		}
+
+		for it.cur.Next() {
+			t0, _ := it.cur.At()
+			if t0 > it.maxt || it.cur.Err() != nil {
+				return false
+			}
+			if t0 >= t {
+				return true
+			}
 		}
 	}
 	return false
 }
 
 func (it *chunkSeriesIterator) At() (t int64, v float64) {
+	// Should it panic if it.cur.Err() != nil || it.i >= len(it.chunks)?
```

Comment:
This comment can be removed.

Reply:
Right, it is now OK to call this at any time. Done.